RunEngine Widget

teddyrendahl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Description
<!--- Describe your changes in detail -->
Basic Widget level control of the RunEngine. This includes:

* A colored label that indicates the current state of the Engine
* A dropdown QComboBox that presents the acceptable actions for the current engine state.
* All performed with signals and slots so other widgets can connect to RunEngine information without writing custom Python code.

### Remaining Questions
* Unfortunately if you expect plans to be run multiple times you need to provide a function that returns a plan not just a plan. This is because after the first run the generator will be exhausted. I couldn't quickly think of anyway around this but open to suggestions right now this looks likes:

```python
EngineWidget(plan=lambda: count(...))
```

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Closes #46 

## Screenshots
![engine](https://user-images.githubusercontent.com/25753048/37112686-002bb232-21f8-11e8-8b8a-73a40c499c3d.gif)

teddyrendahl commented:
This should be re-evaluated in light of recent discussion about `RunEngine` as a `QObject`

ZLLentz commented:
I'm 50/50 on how the widget is set up. I like that it takes two clicks to start the `RunEngine`, I dislike that it takes two clicks to stop it. I'm not sold on the combo box, even though it's future-proof and makes the display very general.

Aesthetically it reminds me a lot of the lcls1 gui.

ZLLentz commented:
I think the widget is fine as is for a first pass, but we'll want to think about how to optimize the layout. I'm not going to tell you how to do this because I don't have any real suggestions on how to do it better.

teddyrendahl commented:
No I need feedback! I'm by in no means in love with the combobox. I might wait to merge and have a RunEngine `QObject`

teddyrendahl commented:
Should it be a stop button or a pause?

ZLLentz commented:
I suppose it should be pause, and then you can evaluate whether to resume or abort with the combobox.

klauer commented:
Agreed

ZLLentz commented:
Also, you should be able to use `itertools.tee` to run plans a second time, so you don't have to pass the empty lambda into the constructor.

klauer commented on typhon/engine.py:
```diff
@@ -0,0 +1,226 @@
+    Parameters
+    ----------
+    engine : RunEngine, optional
+        The underlying RunEngine object. A basic version wil be instatiated if
```
instatiated = in-satiated + instantiated? ;)

klauer commented on typhon/engine.py:
```diff
@@ -0,0 +1,226 @@
+        self._plan = plan
+        # Register a new call command
+        self.command_registry['Start'] = (lambda x:
+                                          RunEngine.__call__(x, self.plan()))
```
If I understand this correctly, you have to touch the right properties in the right order (engine -> plan), which modifies the command_registry dict, making up a new function to call the plan when 'Start' is selected.

This seems a bit over-complicated...  How about instead registering 'Start' as an instance method on the widget? Failing that, special-casing in 'command' below?

teddyrendahl replied:
Interesting. I'll think about this. I also just realized that this setup enforces an order of operations in setting the plan and the engine. This absolutely needs to be corrected.

teddyrendahl commented:
Why is it desirable to have start be two clicks?

What if the QCombobox switches to a pushbutton when there is only one option?

ZLLentz commented:
I don't want to accidentally start a plan, which is much more likely with a single click. This is an issue with `skywalker` where there is a chance you'll run a complicated procedure that wasn't desired when you really just wanted to show off your cool screen.

teddyrendahl commented:
Still have some documentation to do but I reworked this into a better state.

### Major Changes
* The `QComboBox` turns into a `QPushButton` when idle and running. It doesn't really make sense to me to have a `QComboBox` when there is only one option. (I am ignoring that you can halt  the RunEngine, you have to pause -> halt). This ignores @ZLLentz double click request. I think he has a valid point but this requires some thought and probably some use before I can think of a solution.
* All the RunEngine signals and slots are included on a new subclassed `QRunEngine`. This was inspired by some conversations with @tacaswell while he was here. Maybe this is something that belongs in `bluesky`? 
* Fixed the order of operations issue that @klauer pointed out. The `plan_creator` function is attached directly to `QRunEngine` so there is no confusion there.  There is still some kind of strange behavior where you make the widget with an engine and a plan, then swap the RunEngine, you won't have the same plan attached. I thought about automatically attaching the stored plan to the new engine, I just figured that I would rather have people explicitly add plans again instead of surprising folks by running a plan they didn't expect

### Screenshot
![engine_v2](https://user-images.githubusercontent.com/25753048/38327662-a36d4d50-37fd-11e8-99ac-fed66a6f3093.gif)

tacaswell commented on typhon/engine.py:
```diff
@@ -0,0 +1,227 @@
+        self.state_changed.emit(state, old_state)
+
+    @pyqtSlot()
+    def start(self):
```
maybe require that the start slot take in a plan? That moves plan selection logic out to the containing widget and lets you do things like a drop-down list, full UIs for plans or a text box to just type the plan in.

teddyrendahl replied:
👍 to this idea. This simplifies the concerns about attaching plan. It will require a little more work for users though. It probably means most applications will have to have do something like:

```python
but  = QPushButton('My Plan')
but.clicked.connect(functools.partial(QRE.start, my_plan()))
```

or in your example look up the plan from a dictionary based on a string. This is maybe slightly awkward but a lot more flexible. I can always wrap this up in a "PlanButton" or something. Also is there a point to having this by a `pyqtSlot`, can't imagine a lot of buttons will be emitting plans. We probably want to enforce this is a one to one relationship.

tacaswell replied:
plans are all standard python objects, I don't see too much trouble passing live generators across a slot (the c++ layer should just pass it along as an opaque object).

tacaswell commented:
and :+1: this should be in bluesky, I suggest a new top-level module (which might eventually turn into a sub-package) `qt.py`.

tacaswell commented:
See https://github.com/nsls-ii/bluesky-ui for where related effort should land.
